exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -404,21 +420,29 @@ public static Builder builder() {
 
     private EmulatedNodeType nodeType = EmulatedNodeType.VALIDATOR;
     private short validatorCount = 1;
-    private final List<Class<? extends ServiceModule>> services = new ArrayList<>();
-    private TimeProvider timeProvider;
+    private Multimap<ServiceArtifactId, ServiceSpec> services = ArrayListMultimap.create();
+    private HashMap<ServiceArtifactId, String> serviceArtifactFilenames = new HashMap<>();
+    private Path artifactsDirectory;
```

Comment:
Why don't we set defaults for  artifactsDirectory and serverPort here as
    with the rest of the arguments (nodeType, valCount) instead of passing optionals to the constructor? The default port is already documented in the builder setter.

Reply:
Sure, but what would be a good option for the default of `artifactsDirectory`?

Reply:
If there is none — require it to be set, because the current code uses an empty string.

Reply:
I didn't require it to be set in case if TestKit has no user services. But as it's probably just us needing a TestKit with no services, we might as well make `artifactsDirectory` required.